apollo-gradle-plugin/src/main/kotlin/com/apollographql/apollo/gradle/internal/DefaultCompilationUnit.kt

```diff
@@ -65,29 +66,38 @@ abstract class DefaultCompilationUnit @Inject constructor(
     }
   }
 
-  fun setSourcesIfNeeded(graphqlSourceDirectorySet: SourceDirectorySet, schemaFile: RegularFileProperty) {
-    if (graphqlSourceDirectorySet.srcDirs.isEmpty()) {
-      if (schemaFile.isPresent) {
-        graphqlSourceDirectorySet.srcDir(schemaFile.asFile.get().parent)
-      } else {
+  fun setSourcesIfNeeded(sourceDirectorySet: SourceDirectorySet, schemaFile: RegularFileProperty) {
+    if (sourceDirectorySet.srcDirs.isEmpty()) {
+      sourceDirectorySet.findSources(schemaFile)
+    }
+
+    if (!schemaFile.isPresent) {
+      schemaFile.set { resolveSchema(sourceDirectorySet) }
+    }
+  }
+
+  private fun SourceDirectorySet.findSources(schemaFile: RegularFileProperty) {
+    when {
+      apolloVariant.isTest -> srcDirFromVariant(apolloVariant)
```

Comment:
Starting from suggestions in the IDE, I ended up refactoring this to make it easier to follow. This is the only behavior change.

I recommend reviewing this in "split" view.